snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/tracker/TrackerWebViewInterfaceV2Test.kt

```diff
@@ -0,0 +1,195 @@
+ */
+
+package com.snowplowanalytics.snowplow.tracker
+
+import android.content.Context
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.platform.app.InstrumentationRegistry
+import com.snowplowanalytics.core.constants.Parameters
+import com.snowplowanalytics.core.constants.TrackerConstants
+import com.snowplowanalytics.core.emitter.Executor
+import com.snowplowanalytics.core.tracker.TrackerWebViewInterfaceV2
+import com.snowplowanalytics.snowplow.Snowplow.createTracker
+import com.snowplowanalytics.snowplow.Snowplow.removeAllTrackers
+import com.snowplowanalytics.snowplow.configuration.NetworkConfiguration
+import com.snowplowanalytics.snowplow.configuration.TrackerConfiguration
+import com.snowplowanalytics.snowplow.controller.TrackerController
+import com.snowplowanalytics.snowplow.network.HttpMethod
+import com.snowplowanalytics.snowplow.util.EventSink
+import org.json.JSONException
+import org.json.JSONObject
+import org.junit.After
+import org.junit.Assert.*
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class TrackerWebViewInterfaceV2Test {
+    private var webInterface: TrackerWebViewInterfaceV2? = null
+
+    @Before
+    fun setUp() {
+        webInterface = TrackerWebViewInterfaceV2()
+    }
+
+    @After
+    fun tearDown() {
+        removeAllTrackers()
+        Executor.shutdown()
+    }
+
+    @Test
+    @Throws(JSONException::class, InterruptedException::class)
+    fun tracksEventWithAllOptions() {
+        val networkConnection = MockNetworkConnection(HttpMethod.GET, 200)
+        createTracker(
+            context,
+            "ns${Math.random()}",
+            NetworkConfiguration(networkConnection),
+            TrackerConfiguration("appId").base64encoding(false)
+        )
+        
+        val data = "{\"schema\":\"iglu:etc\",\"data\":{\"key\":\"val\"}}"
+        val atomic = "{\"eventName\":\"ue\",\"trackerVersion\":\"webview\"," +
+                "\"useragent\":\"Chrome\",\"pageUrl\":\"http://snowplow.com\"," +
+                "\"pageTitle\":\"Snowplow\",\"referrer\":\"http://google.com\"," +
+                "\"pingXOffsetMin\":10,\"pingXOffsetMax\":20,\"pingYOffsetMin\":30," +
+                "\"pingYOffsetMax\":40,\"category\":\"cat\",\"action\":\"act\"," +
+                "\"property\":\"prop\",\"label\":\"lbl\",\"value\":10.0}"
+        
+        webInterface!!.trackWebViewEvent(
+            selfDescribingEventData = data,
+            atomicProperties = atomic
+        )
+
+        var i = 0
+        while (i < 10 && networkConnection.countRequests() == 0) {
+            Thread.sleep(1000)
+            i++
+        }
+
+        assertEquals(1, networkConnection.countRequests())
+
+        val request = networkConnection.allRequests[0]
+        val payload = request.payload.map
+
+        assertEquals("ue", payload[Parameters.EVENT])
```

Comment:
Can we add a test case for a structured or page view event that their event name is followed? I have an intuition that it will always be set to `ue` because of line 583 in `Tracker.kt`.

I might be wrong and the correct event name is assigned in the end, but a test case never hurts.

Reply:
it's assigned `ue` in Tracker in line 583, that method is called on L528, but then after that on L536 the `TrackerEvent` method `wrapPropertiesToPayload()` is called, where the webview atomic properties are added to the payload. So it ends up with the correct one.

I've updated the first test to use a different event name, to avoid adding another NetworkConnection test.